### ISSUE TITLE
Bluetooth Keyboard Language Layout Compatibility

### ISSUE THREAD

**Issue by dean-futo** [futo-org/android-keyboard#804]
### What happened?

A user is experiencing an issue where connecting a Bluetooth keyboard with a non-English language layout (such as German) automatically switches the layout to English when keyboard is set as the default input method, despite the selected language settings. The physical keyboard layout should align with the chosen language layout, so multilingual typing doesn't cause any involuntary layout changes.

### Keyboard Version

0.1.23

**Comment by p0k33m0n** [comment]
This problem has been with every Android keyboard for many years - the physical keyboard is often hardcoded and that's it. The only developer I found that they can do something about it is Yandex Keyboard (it's a new thing, so maybe in half a year something will come of it).

**Comment by o101010** [comment]
I disagree with p0k33m0n. I use a physical Logitech K400 Plus keyboard with french AZERY layout on my phone. It's corectly detected by Android and I can choose the language in the settings.

I have no probleme when the stock AOSP keyboard (with french language) is selected.
AOSP and english use the wrong qwerty layout.
But with Futo, my physical keyboard is wrongly map as qwerty (with english and french language).

**Comment by p0k33m0n** [comment]
> I have no probleme when the stock AOSP keyboard (with french language) is selected. AOSP and english use the wrong qwerty layout. But with Futo, my physical keyboard is wrongly map as qwerty (with english and french language).

You don't distinguish between a physical keyboard built into a device (e.g. KEYone) and an external keyboard.

**Comment by o101010** [comment]
I have difficulties to undertand the purpose of your message.

This issue mentions bluetooth keyboard, so I suppose external keyboard. 
My keyboard is not bluetooth but usb. I assume the behavior is the same between bt and sub.

My external keyboard is french AZERTY layout.
From my tests : 
 - AOSP keyboard set with english language => usb keyboard is map with qwerty layout
 - AOSP keyboard set with french language => usb keyboard is map with french azerty layout ✅
 - Futo keyboard set with english language => usb keyboard is map with qwerty layout
 - Futo keyboard set with french language => usb keyboard is map with qwerty layout ❌

**Comment by p0k33m0n** [comment]
> I have difficulties to undertand the purpose of your message.

This is for sure.

**Comment by bitboxer** [comment]
I have the same issue on my Xiaomi Pad with external keyboard. When using Gboard it correctly sets the language to De when attaching the keyboard. With Futo the keyboard is english and z/y is switched

**Comment by parametalol** [comment]
Using my Logitech USB keyboard on Android 17, when FUTO is set as the default keyboard I cannot switch layouts as only one layout can be selected.

With Google keyboard I can switch layouts: the system setting for the switch is Ctrl+Space. It also shows a little widget.

**Comment by maninthemiddle01** [comment]
Hey everyone, I’ve been digging into this and wanted to share what I found. I agree with @o101010 and @bitboxer—this isn’t a hard Android limitation. AOSP and Gboard handle external keyboards correctly by respecting the active IME language, so maybe FUTO could do the same.

I asked an LLM which gave me this, so maybe this could help?

**What’s actually happening under the hood:**
The mismatch comes from `java/src/org/futo/inputmethod/event/HardwareKeyboardEventDecoder.java`. Right now, it decodes physical key events using `KeyEvent.getUnicodeChar()`, which pulls characters based on Android’s system-level hardware keyboard layout (usually defaults to US English). It completely ignores the language/subtype you’ve selected in FUTO. There’s even a `TODO` in the constructor that acknowledges this:
```java
// TODO: get the layout for this hardware keyboard
```

When you press a key on a Bluetooth/USB keyboard, `LatinIMELegacy.onKeyDown()` passes the event to the decoder without any context about the current input language. So even if your spacebar says `Deutsch` or `Français`, the physical keys are still mapped using the system default (QWERTY).

**How we could fix it:**
1. **Map keys using the active IME subtype/locale** *(Recommended)*: Pass the current `RichInputMethodSubtype` or `Locale` to the decoder and use FUTO’s existing keyboard layout sets to translate key codes. This would make the physical keyboard behave exactly like the on-screen one, and language switching would work seamlessly.
2. **Use Android’s `KeyCharacterMap` API**: Query the connected device’s actual layout via `KeyCharacterMap.load(deviceId)` and apply it dynamically.
3. **Add a user preference**: Let users manually select their physical keyboard layout if auto-detection isn’t reliable on certain devices.
